### ISSUE TITLE
If all status are set to "visible on front", order confirmation success page throws errors.

### ISSUE THREAD

**Issue by shockwavemk** [magento/magento2#6111]
### Preconditions
1. Magento version: 2.1.0
### Steps to reproduce
1. Go to settings
   
   > Stores > Settings > Order Status
2. Configure all status in magento configuration as visible on front. 
3. Add a product to cart.
4. Go to checkout and proceed to order confirmation.
### Expected result

Order confirmation page is shown.
### Actual result

``` html

Notice: Undefined offset: 0 in /var/www/html/vendor/magento/module-sales/Model/Order/Config.php on line 257

```
### Reason

> \Magento\Sales\Model\Order\Config

``` php
/**
     * Get existing order statuses
     * Visible or invisible on frontend according to passed param
     *
     * @param bool $visibility
     * @return array
     */
    protected function _getStatuses($visibility)
    {
        if ($this->statuses == null) {
            foreach ($this->_getCollection() as $item) {
                $visible = (bool) $item->getData('visible_on_front');
                $this->statuses[$visible][] = $item->getData('status');
            }
        }
        return $this->statuses[(bool) $visibility];
    }
```

Additional problem: the $statuses array is defined as private:

``` php
/**
     * Statuses per state array
     *
     * @var array
     */
    protected $stateStatuses;

    /**
     * @var array
     */
    private $statuses;

    /**
     * @var Status
     */
    protected $orderStatusFactory;
```
### Proposed solution

``` php
/**
 * Order configuration model
 */
class Config extends \Magento\Sales\Model\Order\Config
{
    protected $overridePrivateStatuses;

    /**
     * Get existing order statuses
     * Visible or invisible on frontend according to passed param
     *
     * @param bool $visibility
     * @return array
     */
    protected function _getStatuses($visibility)
    {
        if ($this->overridePrivateStatuses === null) {
            foreach ($this->_getCollection() as $item) {
                $visible = (int) $item->getData('visible_on_front');
                $this->overridePrivateStatuses[$visible][] = $item->getData('status');
            }
        }

        if(array_key_exists((int) $visibility, $this->overridePrivateStatuses)) {
            return $this->overridePrivateStatuses[(int) $visibility];
        }

        return [];
    }
}
```

**Comment by AVoskoboinikov** [comment]
Hello

Thank you for your request, I've created ticket (#MAGETWO-57312) in our internal Jira to fix this.

**Comment by gerryka** [comment]
Can we have an update on this, I just stumbled upon this exact problem on v2.1.4.

**Comment by AVoskoboinikov** [comment]
Fixed in https://github.com/magento/magento2ce/pull/1155

**Comment by korostii** [comment]
Hi @AzVo, your link gives 404 for a non-Magento user (seems to lead to a private repository).
Could you please provide a link to a corresponding commit here on magento/magento2 repo or version in which this bugfix was released?

Thanks in advance.

**Comment by andimov** [comment]
@korostii please, use this link instead https://github.com/magento/magento2/commit/f797367089783b626920fe8daaad6190a5600581

**Comment by korostii** [comment]
Hi @andimov, thank you very much for this link. As I see, it's only merged into the `develop` branch so far.

Wouldn't it be reasonable to keep the issue open until the fix gets into a 2.1.x release?
